### PR TITLE
Add Skill Memory baseline for continual learning

### PR DESCRIPTION
**Pull request by kobros-tech**
## Summary

Adds an experimental Skill Memory baseline for continual learning.

## What changed

* Adds a bounded, immutable memory of acquired model states.
* Probes stored skills on new training data and selects **REUSE**, **CLONE**, or **SCRATCH**.
* Handles model-state restoration and optimizer reset.
* Adds a Permuted MNIST experiment.
* Adds focused unit tests for skill-state storage.
* Does not add an unvalidated benchmark target result.

The implementation is intentionally self-contained so it can be evaluated within the existing continual-learning-baselines workflow.

### CODE REVIEW

**Comment by kobros-tech** [comment]
@AndreaCossu 

I am looking forward to get this contribution reviewed

**Comment by kobros-tech** [comment]
@tachyonicClock 
@GeremiaPompei

**Comment by kobros-tech** [comment]
I made a substantial rewrite of the current implementation here, and I initially missed an important part of the evaluation process.

I now have an alpha version of the strategy divided into two separate operations:

1. **Continual-learning operation** — under the `cl/` directory.
2. **Evaluation operation** — under the `evaluation/` directory.

The most important design change is that the evaluation is now a **pure ML process**. The model responsible for continual learning and the model responsible for evaluation have separate roles.

The first model focuses only on continual learning. It learns each class independently using the Skill Memory mechanism (`REUSE` / `SCRATCH`). There is no task ID, experience ID, class index, or evaluation label used to tell it which skill to use.

The second model is an independent evaluation model. Its only purpose is to learn how to recognize the classes that have been learned by the CL model.

Conceptually:

```text
CL model:
    input x → learns class-specific knowledge

Evaluation model:
    learned class representation → y
```

During evaluation, the input is presented anonymously to the evaluation model. The evaluation model must predict the corresponding `y` without being given the class identity, task ID, experience ID, or Skill Memory index.

So where does `y` come from?

`y` represents the class identity that the CL model originally learned. The evaluation model is trained on the accumulated class information produced by the CL process and learns to associate each learned class with its corresponding `y`.

This separation is important because the CL model operates **per class**, using `REUSE` or `SCRATCH`, but no class/skill index is allowed to leak into the evaluation process.

A useful analogy is how we learn in the real world:

* The machine encounters something new — let's call it **Foo**.
* It learns the characteristics of Foo.
* Later it encounters **Bar** and **Baz** and learns them as well.
* I do not tell the machine which internal index it should use for Foo, Bar, or Baz.
* The evaluation model must learn to distinguish them from their learned representations.

In this setup, the `y` values are simply the machine-learning representation of those names.

So the fundamental separation is:

```text
              ┌──────────────────────┐
              │   CL / Skill Memory  │
              │  REUSE / SCRATCH     │
              │  Learns continuously │
              └──────────┬───────────┘
                         │
                         ▼
              ┌──────────────────────┐
              │    Evaluation ML    │
              │ Learns to recognize │
              │      each class     │
              └──────────┬───────────┘
                         │
                         ▼
                    predicted y
```

The goal is therefore to evaluate whether the knowledge learned by the CL process can be recognized later **without giving the evaluator the class/task/skill identity explicitly**.

**Comment by kobros-tech** [comment]
I develop the strategy code in this repo package:

https://github.com/kobros-tech/cl_strategy

and I compare each approach to replay in ocl_survey here:

https://github.com/kobros-tech/ocl_survey/pull/1

here is just the first positive progress (concerning forgetting for now)

accuracy_plot

forgetting_plot
